hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/index/columnstats/ColumnStatsIndexer.java

```diff
@@ -0,0 +1,667 @@
+      // this can only happen if meta fields are disabled and cols to index is not explicitly overridden.
+      return Collections.singletonList(InitialIndexPartitionData.of(
+          numFileGroup, COLUMN_STATS.getPartitionPath(), engineContext.emptyHoodieData()));
+    }
+
+    LOG.info("Indexing {} columns for column stats index", columnsToIndex.get().size());
+
+    // during initialization, we need stats for base and log files.
+    HoodieData<HoodieRecord> records = convertFilesToColumnStatsRecords(
+        engineContext, Collections.emptyMap(), partitionToFilesMap, dataTableMetaClient,
+        dataTableWriteConfig.getMetadataConfig(),
+        dataTableWriteConfig.getColumnStatsIndexParallelism(),
+        dataTableWriteConfig.getMetadataConfig().getMaxReaderBufferSize(),
+        columnsToIndex.get());
+
+    return Collections.singletonList(InitialIndexPartitionData.of(
+        numFileGroup, COLUMN_STATS.getPartitionPath(), records));
+  }
+
+  @Override
+  public List<IndexPartitionData> update(String instantTime,
+                                         HoodieBackedTableMetadata tableMetadata,
+                                         Lazy<HoodieTableFileSystemView> lazyFileSystemView, HoodieCommitMetadata commitMetadata) {
+    return convertMetadataToColumnStatsRecords(
+        commitMetadata, engineContext,
+        dataTableMetaClient, dataTableWriteConfig.getMetadataConfig(),
+        Option.of(dataTableWriteConfig.getRecordMerger().getRecordType()));
+  }
+
+  @Override
+  public List<IndexPartitionData> clean(String instantTime, HoodieCleanMetadata cleanMetadata) {
+    Option<HoodieData<HoodieRecord>> recordOption = convertMetadataToColumnStatsRecords(cleanMetadata, engineContext,
+        dataTableMetaClient, dataTableWriteConfig.getMetadataConfig(),
+        Option.of(dataTableWriteConfig.getRecordMerger().getRecordType()));
+    return recordOption.isPresent()
+        ? Collections.singletonList(IndexPartitionData.of(COLUMN_STATS.getPartitionPath(), recordOption.get()))
+        : Collections.emptyList();
+  }
+
+  @Override
+  public void updateTableConfig() {
+    // TODO(yihua): though this engine-independent, only Spark has implemented before this PR.
+    //  Revisit to make sure if that's intentional
+    HoodieIndexDefinition indexDefinition = HoodieIndexDefinition.newBuilder()
+        .withIndexName(PARTITION_NAME_COLUMN_STATS)
+        .withIndexType(PARTITION_NAME_COLUMN_STATS)
+        .withIndexFunction(PARTITION_NAME_COLUMN_STATS)
+        .withSourceFields(columnsToIndex.get())
+        .withIndexOptions(Collections.EMPTY_MAP)
+        .build();
+    LOG.info("Registering Or Updating the index {}", PARTITION_NAME_COLUMN_STATS);
+    register(dataTableMetaClient, indexDefinition);
+  }
+
+  /**
+   * Convert added and deleted action metadata to column stats index records.
+   */
+  public static HoodieData<HoodieRecord> convertFilesToColumnStatsRecords(HoodieEngineContext engineContext,
```

Comment:
Can we include methods like this (converting files/records to index recors) in the interface as well?